File: medium/0238_product_of_array_except_self/solution.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        nums_len = len(nums)

        prefix_sums = [1, nums[0]]
        postfix_sums = [nums[-1]] * (nums_len - 1) + [1]

        for i in range(1, nums_len - 1):
            prefix_sums.append(prefix_sums[i] * nums[i])

        for i in range(nums_len - 2, 0, -1):
            postfix_sums[i - 1] = postfix_sums[i] * nums[i]

        return [prefix_sums[i] * postfix_sums[i] for i in range(nums_len)]

File: medium/0238_product_of_array_except_self/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<int> productExceptSelf(std::vector<int> &nums) {
    const size_t numsLen = nums.size();

    std::vector<int> numProducts(numsLen, 1);

    int prefix = 1;
    for (size_t i = 0; i < numsLen; ++i) {
      numProducts[i] = prefix;
      prefix *= nums[i];
    }

    int suffix = 1;
    for (int i = numsLen - 1; i > -1; --i) {
      numProducts[i] *= suffix;
      suffix *= nums[i];
    }

    return numProducts;
  }
};
